cmd/az-log/cmd/common.go

```diff
@@ -0,0 +1,241 @@
+	if err != nil {
+		panic(err.Error())
+	}
+	return clientsetK8s
+}
+
+func GetLogsByAzDriverPod(clientsetK8s kubernetes.Interface, podName string, container string, volumes []string,
+	nodes []string, requestIds []string, since string, sinceTime string, isFollow bool, isPrevious bool) {
+
+	v1PodLogOptions := v1.PodLogOptions{
+		Container: container,
+		Follow:    isFollow,
+	}
+
+	// If since/sinceTime is specified, data type conversion and set up PodLogOptions
+	if since != "" {
+		d, err := time.ParseDuration(since)
+		if err != nil {
+			fmt.Println(err.Error())
+			os.Exit(0)
+		}
+		timeDuration := int64(d.Seconds())
+		v1PodLogOptions.SinceSeconds = &timeDuration
+	} else if sinceTime != "" {
+		t, err := TimestampFormatValidation(sinceTime)
+		if err != nil {
+			fmt.Println(err.Error())
+			os.Exit(0)
+		}
+		timestamp := metav1.NewTime(t)
+		v1PodLogOptions.SinceTime = &timestamp
+	}
+
+	podLogOptions := make([]v1.PodLogOptions, 0)
+
+	// If logs from previous container is needed
+	if isPrevious {
+		v1PodLogOptions.Previous = true
+		podLogOptions = append(podLogOptions, v1PodLogOptions)
+		v1PodLogOptions.Previous = false
+	}
+
+	podLogOptions = append(podLogOptions, v1PodLogOptions)
+
+	for i := 0; i < len(podLogOptions); i++ {
+		req := clientsetK8s.CoreV1().Pods(consts.ReleaseNamespace).GetLogs(podName, &podLogOptions[i])
+		podLogs, err := req.Stream(context.TODO())
+		if err != nil {
+			panic(err.Error())
+		}
+
+		defer podLogs.Close()
+
+		buf := bufio.NewScanner(podLogs)
+		LogFilter(buf, volumes, nodes, requestIds, "")
+	}
+}
+
+func TimestampFormatValidation(sinceTime string) (time.Time, error) {
+	var t time.Time
+	var err error
+
+	// sinceTime input validation and convert it to time.Time from string
+	if isMatch, _ := regexp.MatchString(RFC3339Format, sinceTime); isMatch {
+		t, err = time.Parse(time.RFC3339, sinceTime)
+		if err != nil {
+			return t, fmt.Errorf("error: %v", err)
+		}
+	} else if isMatch, _ := regexp.MatchString(KlogTimeFormat, sinceTime); isMatch {
+		if isUTC, _ := regexp.MatchString(KlogTimeFormat+`$`, sinceTime); isUTC {
+			sinceTime += "Z"
+		}
+
+		t, err = time.Parse("20060102 15:04:05Z07:00", fmt.Sprint(time.Now().Year())+sinceTime)
+		if err != nil {
+			return t, fmt.Errorf("error: %v", err)
+		}
+
+	} else {
+		return t, fmt.Errorf("\"%v\" is not a valid timestamp format", sinceTime)
+	}
+
+	return t, nil
+}
+
+func LogTimeFilter(log string, sinceTime string) bool {
+	isMatch, _ := regexp.MatchString(KlogTimeFormat, log)
+	return isMatch && len(log) > 21 && log[1:21] >= sinceTime
+}
+
+func LogFilter(buf *bufio.Scanner, volumes []string, nodes []string, requestIds []string, sinceTime string) {
+	var isAfterTime bool
+	if sinceTime == "" {
+		isAfterTime = true
+	} else {
+		isAfterTime = false
+	}
+
+	for buf.Scan() {
+		log := buf.Text()
+
+		if !isAfterTime {
+			isAfterTime = LogTimeFilter(log, sinceTime)
+		}
+
+		if isAfterTime {
+			isPrint := true
+			if len(volumes) > 0 {
```

Comment:
Instead of having multiple conditions to check whether the volumes/nodes/requestIds are present in a line, can we have a single check separated by or/and conditions?

Reply:
For that we could use different functions for each filter, like containsVolumes(volumes..) && containsNodes(nodes..) && containsRequestIds(requestIds..)

Reply:
Does it make sense to just use one function like ContainsAny for volumes/nodes/requestids since they are all checking if a log contains a specific substring?